### PR TITLE
memmove for overlaping memory

### LINKED ISSUE #433: UB: memcpy with overlapping memory in rcutils_array_list_remove when destroying child logger during tear down
This is [test output from mtc with asan](https://github.com/ros-planning/moveit_task_constructor/actions/runs/6731269933/job/18330877384?pr=502). 

```
    [moveit_task_constructor_core-test-move-to-1] ==6030==ERROR: AddressSanitizer: memcpy-param-overlap: memory ranges [0x60300033e200,0x60300033e210) and [0x60300033e208, 0x60300033e218) overlap
    [moveit_task_constructor_core-test-move-to-1]     #0 0x7f8c9d26d1ed in __interceptor_memcpy ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:827
    [moveit_task_constructor_core-test-move-to-1]     #1 0x7f8c9d121322 in rcutils_array_list_remove (/opt/ros/rolling/lib/librcutils.so+0x7322)
    [moveit_task_constructor_core-test-move-to-1]     #2 0x7f8c9d122cbd in rcutils_hash_map_unset (/opt/ros/rolling/lib/librcutils.so+0x8cbd)
    [moveit_task_constructor_core-test-move-to-1]     #3 0x7f8c9a6bd286  (/opt/ros/rolling/lib/librcl.so+0x1a286)
    [moveit_task_constructor_core-test-move-to-1]     #4 0x7f8c9a6c2459 in rcl_logging_rosout_fini_publisher_for_node (/opt/ros/rolling/lib/librcl.so+0x1f459)
    [moveit_task_constructor_core-test-move-to-1]     #5 0x7f8c9c7935cb  (/opt/ros/rolling/lib/librclcpp.so+0x12e5cb)
    [moveit_task_constructor_core-test-move-to-1]     #6 0x7f8c9c798167  (/opt/ros/rolling/lib/librclcpp.so+0x133167)
    [moveit_task_constructor_core-test-move-to-1]     #7 0x7f8c9c7568c9  (/opt/ros/rolling/lib/librclcpp.so+0xf18c9)
    [moveit_task_constructor_core-test-move-to-1]     #8 0x7f8c9c790426 in rclcpp::node_interfaces::NodeBase::~NodeBase() (/opt/ros/rolling/lib/librclcpp.so+0x12b426)
```

My understanding from reading about `memcpy-param-overlap` is that we should use `memmove` when the memory addresses in the source and destination overlap.  Here is the offending line we should change: https://github.com/ros2/rcutils/blob/e276dc1fe5a0e450e53423ec71d57cbe05b129f8/src/array_list.c#L175

### PR DESCRIPTION
**Pull request by tylerjw**
Fixes #433

### CODE REVIEW

**Comment by tylerjw** [comment]
I'm unsure what to do next. I assume some huge CI job needs to run to test this?

**Comment by clalancette** [comment]
> I'm unsure what to do next. I assume some huge CI job needs to run to test this?

Well, the commit needs a `Signed-off-by` line, and I think it would be good to have a test exercising this.  But after that, and review, yes, it needs full CI (which one of the maintainers will run).

**Comment by tylerjw** [comment]
Is there a way to add testing this with asan (as that is how we found this)?

**Comment by clalancette** [comment]
> Is there a way to add testing this with asan (as that is how we found this)?

Theoretically, https://ci.ros2.org/view/nightly/job/nightly_linux_address_sanitizer/ should catch this.  So if there is a test that exercises it, it should show up there.

**Comment by tylerjw** [comment]
Added the signed-off-by line.

I'll try to add a test that exercises this.

**Comment by tylerjw** [comment]
I added a test.

File: src/array_list.c [hunk 1]
```diff
@@ -172,7 +172,7 @@ rcutils_array_list_remove(rcutils_array_list_t * array_list, size_t index)
   if (copy_count > 0) {
     uint8_t * dst_ptr = rcutils_array_list_get_pointer_for_index(array_list, index);
     uint8_t * src_ptr = rcutils_array_list_get_pointer_for_index(array_list, index + 1);
-    memcpy(dst_ptr, src_ptr, array_list->impl->data_size * copy_count);
+    memmove(dst_ptr, src_ptr, array_list->impl->data_size * copy_count);
```
Comment (clalancette) [inline src/array_list.c:175]: I think it is worthwhile to add a comment here saying why we are using a `memmove` instead of a `memcpy`.

**Review by clalancette — CHANGES_REQUESTED** [review]
I tested this locally with enabling asan, and it does show the problem.  So the test seems good.  I left one thing inline that I think we should add a comment to; once that is added I'll run full CI on this.

File: src/array_list.c (hunk 1 above)
Comment (tylerjw) [reply src/array_list.c:179]: Added a comment, let me know if it does a good enough job explaining the use or if you'd like it rephrased in any way?

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19722/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14229/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20448/)

**Review by fujitatomoya — APPROVED** [review]
lgtm :+1:

**Comment by clalancette** [comment]
@Mergifyio backport iron humble

**Pull request merged**